content/posts/monad-expressions-in-csharp.md

```diff
@@ -0,0 +1,131 @@
++++
+title =  "C# expressions that try to be monadic"
+description =  ""
+date =  2021-07-01
+template = "blog/page.html"
+draft =  false
+
+[taxonomies]
+authors = ["Robert Marciniak"]
+#tags =  ["monad", "linq", "async"]
+#categories =  []
++++
+
+# Why we can't have nice things, c# query expression edition
+
+The closest thing to f#'s computational expression is either combination of `async` `await` keywords in contexts that
+do not always translate to asynchronous context or query expression, which does not always translate to querying something.
+
+That's because C# is a corporate enterprise language which tries to smuggle Nice Things through, but in a different,
+more corporate-enterprise-like form - not too abstract, sprinkled with "query SQL from code!"-, "Improve responsiveness without big refactors!"-, or similar seasoning.
+
+## 'what's the underlying type?'
+
+### Case 1
+
+```cs
+from index in Enumerable.Range(1, 5)
+from date in Enumerable.Repeat(DateTime.Today, 5)
+select (index, date) into pair
+select pair.date.AddDays(pair.index);
+```
+
+```fs
+monad {
+    let! index = Enumerable.Range(1, 5)
+    let! date = Enumerable.Repeat(DateTime.Today, 5)
+    let pair = (index, date)
+    yield pair |> (fun (i, d) -> d.AddDays(i |> float))
+}
+```
+
+### Case 2
+
+```cs
+from index in GetIndexAsync()
+from date in Task.FromResult(DateTime.Today)
+select (index, date) into pair
+select pair.date.AddDays(pair.index);
+```
+
+```fs
+monad {
+    let! index = GetIndexAsync()
+    let! date = Task.FromResult(DateTime.Today)
+    let pair = (index, date)
+    yield pair |> (fun (i, d) -> d.AddDays(i |> float))
+}
+```
+
+### Case 3
+```cs
+from index in Option.Some(5)
+from date in Option.None<DateTime>()
+select (index, date) into pair
+select pair.date.AddDays(pair.index);
+```
+
+```cs
+async Option<DateTime> Fun() {
+    var index = await Option.Some(5);
+    var date = await Option.None<DateTime>();
+    var pair = (index, date);
+    return pair.date.AddDays(pair.index)
+}
+```
+
+```fs
+monad {
+    let! index = Some 5
+    let! date = None
+    let pair = (index, date)
+    yield pair |> (fun (i, (d:DateTime)) -> d.AddDays(i |> float))
+}
+```
+## ...converging into the same topic, never reaching it
+
+These c# constructs do exactly the same thing, without any major or unsafe hacks. The main barrier from using these in a monad-like manner is their naming. Despite being perfectly viable in multiple scenarios their naming suggests using them in LINQ-to-SQL and asynchronous code respectively.

```

Comment:
You refer to LINQ-to-SQL which works on IQueryable, but actually showed a version that works on IEnumerables. There is also a version that queries XML. I think the naming does not suggest which version of LINQ you are using.

I think you wanted to express, that naming of LINQ is similar to SQL. Am I correct?